File: sesion5-lab/spring-jwt-attacks/jwt-server/src/main/java/com/example/jwt/service/VulnerableJwtService.java
package com.example.jwt.service;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.Map;

import javax.crypto.SecretKey;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import com.fasterxml.jackson.databind.ObjectMapper;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;

@Service
public class VulnerableJwtService {

    private final String secret;
    private final ObjectMapper mapper = new ObjectMapper();

    public VulnerableJwtService(@Value("${jwt.vulnerable.secret}") String secret) {
        this.secret = secret;
    }

    public String issueToken(String subject, String role) {
        SecretKey key = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
        return Jwts.builder()
                .subject(subject)
                .claim("role", role)
                .issuedAt(new Date())
                .expiration(new Date(System.currentTimeMillis() + 3600_000))
                .signWith(key)
                .compact();
    }

    /**
     * VULN: acepta alg=none (sin firma) y no valida iss/aud.
     */
    public Map<String, Object> validate(String token) {
        String[] parts = token.split("\\.");
        if (parts.length < 2) {
            throw new IllegalArgumentException("Token malformado");
        }

        String headerJson = new String(Base64.getUrlDecoder().decode(parts[0]), StandardCharsets.UTF_8);
        Map<?, ?> header = readJson(headerJson);
        String alg = String.valueOf(header.get("alg"));

        Claims claims;
        if ("none".equalsIgnoreCase(alg)) {
            // VULN critica — acepta tokens sin firma
            String payloadJson = new String(Base64.getUrlDecoder().decode(parts[1]), StandardCharsets.UTF_8);
            claims = Jwts.claims(readJson(payloadJson));
        } else {
            SecretKey key = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
            claims = Jwts.parser()
                    .verifyWith(key)
                    .build()
                    .parseSignedClaims(token)
                    .getPayload();
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("valid", true);
        result.put("algorithm", alg);
        result.put("subject", claims.getSubject());
        result.put("role", claims.get("role"));
        result.put("secretLength", secret.length());
        result.put("issues", new String[] {
                "Acepta algoritmo none",
                "Secreto HMAC debil (" + secret.length() + " chars)",
                "No valida iss ni aud"
        });
        return result;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> readJson(String json) {
        try {
            return mapper.readValue(json, Map.class);
        } catch (Exception ex) {
            throw new IllegalArgumentException("JSON invalido en JWT", ex);
        }
    }
}

File: sesion5-lab/aspnet-jwt-attacks/jwt-demo/Services/VulnerableJwtService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using System.Text.Json;
using Microsoft.IdentityModel.Tokens;

namespace JwtDemo.Services;

public class VulnerableJwtService(IConfiguration config)
{
    private readonly string _secret = config["Jwt:Vulnerable:Secret"] ?? "MiSecreto12";

    public string IssueToken(string subject, string role)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_secret));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var token = new JwtSecurityToken(
            claims: [new Claim("sub", subject), new Claim("role", role)],
            expires: DateTime.UtcNow.AddHours(1),
            signingCredentials: creds);
        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    public object Validate(string token)
    {
        var parts = token.Split('.');
        if (parts.Length < 2)
        {
            throw new ArgumentException("Token malformado");
        }

        var headerJson = Encoding.UTF8.GetString(Base64UrlDecode(parts[0]));
        using var header = JsonDocument.Parse(headerJson);
        var alg = header.RootElement.GetProperty("alg").GetString() ?? "";

        JwtSecurityToken jwt;
        if (alg.Equals("none", StringComparison.OrdinalIgnoreCase))
        {
            var payloadJson = Encoding.UTF8.GetString(Base64UrlDecode(parts[1]));
            using var payload = JsonDocument.Parse(payloadJson);
            var claims = payload.RootElement.EnumerateObject()
                .Select(p => new Claim(p.Name, p.Value.ToString()))
                .ToList();
            jwt = new JwtSecurityToken(claims: claims);
        }
        else
        {
            var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_secret));
            var handler = new JwtSecurityTokenHandler();
            handler.ValidateToken(token, new TokenValidationParameters
            {
                ValidateIssuer = false,
                ValidateAudience = false,
                IssuerSigningKey = key,
                ValidateIssuerSigningKey = true
            }, out var validated);
            jwt = (JwtSecurityToken)validated;
        }

        return new
        {
            valid = true,
            algorithm = alg,
            subject = jwt.Claims.FirstOrDefault(c => c.Type is "sub" or ClaimTypes.NameIdentifier)?.Value,
            role = jwt.Claims.FirstOrDefault(c => c.Type == "role")?.Value,
            secretLength = _secret.Length,
            issues = new[] {
                "Acepta algoritmo none",
                $"Secreto HMAC debil ({_secret.Length} chars)",
                "No valida iss ni aud"
            }
        };
    }

    private static byte[] Base64UrlDecode(string input)
    {
        var padded = input.Replace('-', '+').Replace('_', '/');
        switch (padded.Length % 4)
        {
            case 2: padded += "=="; break;
            case 3: padded += "="; break;
        }
        return Convert.FromBase64String(padded);
    }
}
